File: ws/rest/runner.js
/*
 * Runner RESTful web services
 *
 * This is a RESTful web services that execute APPRIS pipeline.
 */

/**
 * @api {post} /run run > Submit a job with the specified parameters
 * @apiName run
 * @apiGroup Runner
 * @apiVersion 0.1.0
 *
 * @apiParam {String} species 	Species name (homo_sapiens, hsap, mus_musculus, mmus, danio_rerio, drer).
 * @apiParam {String} sequences	Fasta protein sequence.
 * @apiParam {String} id 		Gene identifier/name.
 * @apiParam {String} e_version	Ensembl version.
 * @apiParam {String} methods 	Methods to execute (appris,firestar,matador3d,spade,corsair,crash,thump).
 * @apiParam {String} outfile 	Output file name.
 * @apiParam {String} format 	Output format.
 * 
 * 		raw: result of every method in text plain format
 * 		json: completed result of every method in json format
 *		tsv: highlight results of every method in tabular
 *		gtf: genomic data of every method in gtf format
 * 		bed: genomic data of every method in bed format 
 *
 * @apiSuccess {String} jobid A plain text document containing the job identifier.
 *
 * @apiExample Example usage:
 * http://apprisws.bioinfo.cnio.es/ws/rest/runner/run
 *	POST
 *	{
 *	'species' => 'Homo sapiens',
 *	'id' => 'ENSG00000099999',
 *	'e_version' => '70',
 *	'methods' => 'firestar,matador3d,corsair,spade,thump,crash,appris',
  *	'format' => 'raw',
 *	}; 
 *
 * @apiSuccessExample Success-Response:
 *     HTTP/1.1 200 OK
 *     c5fcac3d75a7e6d56b753c2265a73ad1
 *
 * @apiError BadRequest The request is wrong.
 * @apiError JobNotFound The <code>id</code> of the Job was not found.
 * @apiError MethodNotAllowed The <code>method</code> is not allowed.
 *
 * @apiErrorExample Error-Response:
 *     HTTP/1.1 400 Bad Request The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
 *     HTTP/1.1 404 Not Found The server has not found anything matching the Request-URI.
 *     HTTP/1.1 405 Method Not Allowed: The parameter XXX is not allowed. The parameter must be: YYY.
 *
 */
 
 /**
 * @api {get} /status/:jobid status > Get the status of a submitted job.
 * @apiName status
 * @apiGroup Runner
 * @apiVersion 0.1.0
 *
 * @apiParam {String} jobid 	Job identifier.
 *
 * @apiSuccess {String} status A plain text document containing the job status. The values for the status are:
 * 
 * RUNNING: the job is currently being processed.
 * FINISHED: job has finished, and the results can then be retrieved.
 * ERROR: an error occurred attempting to get the job status.
 * FAILURE: the job failed.
 * NOT_FOUND: the job cannot be found.
 *
 * @apiExample Example usage:
 * http://apprisws.bioinfo.cnio.es/ws/rest/runner/status/2eb1bb92f32a0384b1b417cb7c9a02ff
 *
 * @apiSuccessExample Success-Response:
 *     HTTP/1.1 200 OK
 *     	RUNNING
 *
 * @apiError BadRequest The request is wrong.
 * @apiError JobNotFound The <code>id</code> of the Job was not found.
 * @apiError MethodNotAllowed The <code>method</code> is not allowed.
 *
 * @apiErrorExample Error-Response:
 *     HTTP/1.1 400 Bad Request The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
 *     HTTP/1.1 404 Not Found The server has not found anything matching the Request-URI.
 *     HTTP/1.1 405 Method Not Allowed: The parameter XXX is not allowed. The parameter must be: YYY.
 *
 */
 
/**
 * @api {get} /resulttypes/:jobid resulttypes > Get available result types for a finished job.
 * @apiName resulttypes
 * @apiGroup Runner
 * @apiVersion 0.1.0
 *
 * @apiParam {String} jobid 	Job identifier.
 *
 * @apiSuccess {JSON} resulttypes 	An JSON detailing the available result types.
 *
 * @apiExample Example usage:
 * http://apprisws.bioinfo.cnio.es/ws/rest/runner/resulttypes/2eb1bb92f32a0384b1b417cb7c9a02ff
 *
 * @apiSuccessExample Success-Response:
 *     HTTP/1.1 200 OK
 *     	{
 *     	'identifier' => 'appris',
 *     	'fileSuffix' => 'raw',
 *     	'mediaType' => 'text/plain',
 *     	'description' => ''
 *     	};
 *
 * @apiError BadRequest The request is wrong.
 * @apiError JobNotFound The <code>id</code> of the Job was not found.
 * @apiError MethodNotAllowed The <code>method</code> is not allowed.
 *
 * @apiErrorExample Error-Response:
 *     HTTP/1.1 400 Bad Request The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
 *     HTTP/1.1 404 Not Found The server has not found anything matching the Request-URI.
 *     HTTP/1.1 405 Method Not Allowed: The parameter XXX is not allowed. The parameter must be: YYY.
 *
 */

 /**
 * @api {get} /result/:jobid result > Get the job result of the specified type.
 * @apiName result
 * @apiGroup Runner
 * @apiVersion 0.1.0
 *
 * @apiParam {String} jobid 		Job identifier.
 * @apiParam {String} format 		Format result.
 * 
 * 		raw: result of every method in text plain format
 * 		json: completed result of every method in json format
 *		tsv: highlight results of every method in tabular
 *		gtf: genomic data of every method in gtf format
 * 		bed: genomic data of every method in bed format
 * @apiParam {String} headbed 		Add head info into bed file. The values are:
 * 
 * 		no: no head information
 * 		yes: ensGene,ccdsGene,burgeRnaSeqGemMapperAlign add head tracks to bed file
 *		only: print only the head of bed format
 *
 * @apiSuccess {String} result 	A document containing the result in the requested format. The MIME type of the returned document is set according to the format.
 *
 * @apiExample Example usage:
 * http://apprisws.bioinfo.cnio.es/ws/rest/runner/result/c5fcac3d75a7e6d56b753c2265a73ad1?format=tsv
 *
 * @apiSuccessExample Success-Response:
 *     HTTP/1.1 200 OK
 *     == APPRIS > FIRESTAR
 *     == >sequence_id
 *     == residue	amino_acid	ligand	reliability_score (1-->6)
 *     >ENST00000405930
 *     106	C	ZN	4
 *     109	C	ZN	4
 *     120	C	ZN,ZN	6,6
 *     123	C	ZN,ZN	6,4
 *     132	H	ZN	6
 *     134	C	ZN	6
 *
 * @apiError BadRequest The request is wrong.
 * @apiError JobNotFound The <code>id</code> of the Job was not found.
 * @apiError MethodNotAllowed The <code>method</code> is not allowed.
 *
 * @apiErrorExample Error-Response:
 *     HTTP/1.1 400 Bad Request The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
 *     HTTP/1.1 404 Not Found The server has not found anything matching the Request-URI.
 *     HTTP/1.1 405 Method Not Allowed: The parameter XXX is not allowed. The parameter must be: YYY.
 *
 */

  /**
 * @api {get} /result/:jobid/:method result > Get the job result of the specified type filtering by a given method.
 * @apiName result
 * @apiGroup Runner
 * @apiVersion 0.1.0
 *
 * @apiParam {String} jobid 		Job identifier.
 * @apiParam {String} methods 	Methods to execute (appris,firestar,matador3d,spade,corsair,crash,thump).
 * @apiParam {String} format 		Format result.
 * 
 * 		raw: result of every method in text plain format
 * 		json: completed result of every method in json format
 *		tsv: highlight results of every method in tabular
 *		gtf: genomic data of every method in gtf format
 * 		bed: genomic data of every method in bed format
 * @apiParam {String} headbed 		Add head info into bed file. The values are:
 * 
 * 		no: no head information
 * 		yes: ensGene,ccdsGene,burgeRnaSeqGemMapperAlign add head tracks to bed file
 *		only: print only the head of bed format
 *
 * @apiSuccess {String} result 	A document containing the result in the requested format. The MIME type of the returned document is set according to the format.
 *
 * @apiExample Example usage:
 * http://apprisws.bioinfo.cnio.es/ws/rest/runner/result/d2b90a9c5a7bd1b0272c930eccabc3f2/corsair?format=tsv
 *
 * @apiSuccessExample Success-Response:
 *     HTTP/1.1 200 OK
 *		== APPRIS > CORSAIR
 *		== >sequence_id
 *		== nearest_homologue	%ID
 *		>ENST00000405930
 *		Homo sapiens	100.00
 *		>ENST00000436518
 *		>ENST00000320602
 *		>ENST00000334554
 *		Homo sapiens	100.00
 *		Pan troglodytes	96.49
 *		Sus scrofa	94.12
 *		Canis lupus familiaris	93.86
 *		Rattus norvegicus	91.37
 *		Mus musculus	91.37
 *
 * @apiError BadRequest The request is wrong.
 * @apiError JobNotFound The <code>id</code> of the Job was not found.
 * @apiError MethodNotAllowed The <code>method</code> is not allowed.
 *
 * @apiErrorExample Error-Response:
 *     HTTP/1.1 400 Bad Request The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
 *     HTTP/1.1 404 Not Found The server has not found anything matching the Request-URI.
 *     HTTP/1.1 405 Method Not Allowed: The parameter XXX is not allowed. The parameter must be: YYY.
 *
 */

File: ws/rest/runner.pl
#!/usr/bin/perl

use strict;
use warnings;
use CGI;
use JSON;
use Email::Valid;
use Data::Dumper;
use FindBin;
use lib "$FindBin::Bin/lib";
use HTTP qw( print_http_response );
use WSRunner;
#$|=1; # not use buffering

###################
# Global variable #
###################
use vars qw(
	$FORMAT
	$METHODS
	$HEAD_BED
	$ENCODING
);

$FORMAT = $ENV{APPRIS_WSERVER_OUTPUT_FORMAT};
$METHODS = $ENV{APPRIS_WSERVER_PIPELINE_STRUCTURE};
$HEAD_BED = $ENV{APPRIS_WSERVER_HEADBEAD};
$ENCODING = $ENV{APPRIS_WSERVER_OUTPUT_ENCODING};

#####################
# Method prototypes #
#####################
sub main();

#################
# Method bodies #
#################

# Main subroutine
sub main()
{
	# Get input parameters:
	my ($cgi) = new CGI;
	
	# Check input parameters:
	my ($url_path_info) = $cgi->path_info();
	my (@input_parameters) = split('/',$url_path_info);
		
	print_http_response(400)
		unless ( @input_parameters ); # defined path
		
	print_http_response(400)
		if ( scalar(@input_parameters) >= 4 );
	my ($type) = $input_parameters[1];
	my ($jobid) = $input_parameters[2];
	
	print_http_response(405, "The parameter $type is not allowed. The parameter must be: 'run', 'status', 'resulttypes', or 'result'.")
		if ( ($type ne 'run') and ($type ne 'status') and ($type ne 'resulttypes') and ($type ne 'result') );
	
	# get postdata if exits
	my ($pdata) = $cgi->param('POSTDATA');
	my ($postdata) = decode_json($pdata) if ( defined $pdata and ($pdata ne '') );
	
	# Collect parameters for execution:
	my ($params);	
	if ( $type eq 'run' )
	{					
		my ($species) = $cgi->param('species') || $postdata->{'species'} || undef;
		if ( defined $species ) {
			$params->{'species'} = $species;
		}
		else {
			print_http_response(400);
		}
			
		my ($id) = $cgi->param('id') || $postdata->{'id'} || undef;
		my ($e_version) = $cgi->param('e_version') || $postdata->{'e_version'} || undef;
		my ($sequences) = $cgi->param('sequences') || $postdata->{'sequences'} || undef;
	
		if ( defined $id and ($id ne '') and defined $e_version and ($e_version ne '') and !defined $sequences ) {
			$params->{'id'} = $id;
			$params->{'e_version'} = $e_version;
		}
		elsif ( defined $sequences and ($sequences ne '') and !defined $id and !defined $e_version ) {
			$params->{'sequences'} = $sequences;
		}
		else {
			print_http_response(400);
		}
		
		my ($methods) = $cgi->param('methods') || $postdata->{'methods'} || undef;
		if (defined $methods and ($methods ne '')) {
			$methods = lc($methods);		
			if ( 
				($methods =~ /appris/) or
				($methods =~ /firestar/) or
				($methods =~ /matador3d/) or
				($methods =~ /spade/) or
				($methods =~ /corsair/) or
				($methods =~ /crash/) or
				($methods =~ /thump/)
			) {
				$METHODS = $methods;	
			}
			else {
				print_http_response(405, "The parameter $methods is not allowed. The parameter must be: $METHODS.")
			}
		}
		$params->{'methods'} = $METHODS; # by default, we execute all methods
		
		my ($email) = $cgi->param('email') || $postdata->{'email'} || undef;
		if (defined $email and ($email ne '')) {
			if ( Email::Valid->address($email) ) {
				$params->{'email'} = $email;
			}
			else {
				print_http_response(405, "The parameter $email must be valid")
			}
		}	
	}
	elsif ( $type eq 'result' )
	{
		my ($methods) = $cgi->param('methods') || $postdata->{'methods'} || undef;
		if ( defined $methods ) {
			$params->{'methods'} = $methods;
		}
		
		my ($ids) = $cgi->param('ids') || $postdata->{'ids'} || undef;
		if ( defined $ids ) {
			$params->{'ids'} = $ids;
		}
		
		my ($res) = $cgi->param('res') || $postdata->{'res'} || undef;
		if ( defined $res ) {
			$params->{'res'} = $res;
		}

		my ($format) = $cgi->param('format') || $postdata->{'format'} || undef ;
		if (defined $format and ($format ne '')) {
			$format = lc($format);
			if (($format ne '') and ( ($format eq 'raw') or ($format eq 'tsv') or ($format eq 'gtf') or ($format eq 'bed') or ($format eq 'json') ) ) {
				$FORMAT = $format;	
			}
			else {
				print_http_response(405, "The parameter $format is not allowed.")
			}
		}
		$params->{'format'} = $FORMAT;
		
		my ($head) = $cgi->param('headbed') || $postdata->{'headbed'} || undef;
		if (defined $head and ($head ne '') and ($format eq 'bed') ) {
			$head =~ s/\s*//g;
			if ($head ne '' and ( ($head =~ /^yes/) or ($head =~ /^no/) or ($head =~ /^only/) ) ) {
				$HEAD_BED = $head;	
			}
			else {
				print_http_response(405, "The parameter $head is not allowed.")
			}
			$params->{'headbed'} = $HEAD_BED;
		}
		
		my ($email) = $cgi->param('email') || $postdata->{'email'} || undef;
		if (defined $email and ($email ne '')) {
			if ( Email::Valid->address($email) ) {
				$params->{'email'} = $email;
			}
			else {
				print_http_response(405, "The parameter $email must be valid")
			}
		}
	}
	elsif ( $type eq 'resulttypes' )
	{					
		my ($species) = $cgi->param('species') || $postdata->{'species'} || undef;
		if ( defined $species ) {
			$params->{'species'} = $species;
		}
	}	
	
	# Runner object
	my ($runner) = new WSRunner(
								-oper	=> $type,
								-params => $params,
								-jobid	=> $jobid,
	);

	# Print data
	my ($result);
	if ( $type eq 'run' ) {
		$result = $runner->apply_request($type);
	}
	elsif ( $type eq 'status' ) {
		$result = $runner->apply_request($type);
	}
	elsif ( $type eq 'resulttypes' ) {
		$result = $runner->apply_request($type);
	}
	elsif ( $type eq 'result' ) {
		$result = $runner->apply_request($type);
	}
	else {
		print_http_response(404);
	}
		
	# Everything was OK
	if ( defined $result and ($result ne '') ) {
		if ( $FORMAT eq 'json' ) {
			$ENCODING = 'application/json';
		}
		print_http_response(200, $result, $ENCODING);
	}
	else {
		print_http_response(404);
	}	

}

main();


1;


__END__

=head1 NAME

runner

=head1 DESCRIPTION

RESTful web services that execute APPRIS pipeline

=head1 SYNOPSIS

=head2 Required arguments:

=head1 EXAMPLES

=head1 AUTHOR

Jose Manuel Rodriguez Carrasco -jmrodriguez@cnio.es- (INB-GN2,CNIO)

=cut
